File: examples/python/stack.py
# invariant: self.n >= 0
class Stack:
    # ensures: self.n == 0
    def __init__(self):
        self.n = 0
    # ensures: self.n == old(self.n) + 1
    def push(self, x):
        self.n += 1
    # requires: self.n > 0
    # ensures: self.n == old(self.n) - 1
    def pop(self):
        self.n -= 1
        return 0

File: examples/rust/stack.rs
pub struct Stack { n: i32 }

// invariant: n >= 0
impl Stack {
    // ensures: n == 0
    pub fn new() -> Self { Stack { n: 0 } }
    // ensures: n == old(n) + 1
    pub fn push(&mut self, _x: i32) { self.n += 1; }
    // requires: n > 0
    // ensures: n == old(n) - 1
    pub fn pop(&mut self) -> i32 { self.n -= 1; 0 }
}
